### PR TITLE
Add an option to disable fail fast and failure accural for Consul

### PR DESCRIPTION
**Pull request by Ashald**

### CODE REVIEW

**Comment by Ashald** [comment]
Not sure whether docs/changelog are good enough though. :/

File: consul/src/main/scala/io/buoyant/consul/utils.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import com.twitter.util.Monitor
+import io.buoyant.config.types.Port
+import java.net.InetSocketAddress
+
+
+object utils {
+
+  implicit class RichConsulClient(val client: Http.Client) {
+
+    val monitor = Monitor.mk {
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: can we move this to the `utils` object or the `RichConsulClient` companion object?

File: consul/src/main/scala/io/buoyant/consul/utils.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import com.twitter.finagle.http.{Request, Response}
+import com.twitter.finagle.{Failure, Http, Service, SimpleFilter}
+import com.twitter.util.Monitor
+import io.buoyant.config.types.Port
+import java.net.InetSocketAddress
+
+
+object utils {
+
+  implicit class RichConsulClient(val client: Http.Client) {
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: Make this a value class by adding `extends AnyVal`

File: consul/src/main/scala/io/buoyant/consul/utils.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    val monitor = Monitor.mk {
+      case e: Failure if e.isFlagged(Failure.Interrupted) => true
+    }
+
+    def failFast(enabled: Option[Boolean]) =
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: explicit return types on public methods, please

File: CHANGES.md [hunk 4]
```diff
@@ -17,6 +17,7 @@
   is deployed as a k8s daemonset.
 * Add `consistencyMode` option to `io.l5d.consul` namer
 * Add `readConsistencyMode` and `writeConsistencyMode` options to `io.l5d.consul` dtab storage
+* Add `failFast` (defaults to `false`) to consul namer and dtab storage options
```
Comment (adleong) [inline CHANGES.md]: Hmmm... this is a change in the default behavior so I might word it to emphasize that.  Something like "failFast is now disabled by default for the consul namer and dtab storage but can be enabled with the `failFast`option"

File: namerd/docs/storage.md [hunk 5]
```diff
@@ -98,4 +98,5 @@ token | no auth | The auth token to use when making API calls.
 datacenter | uses agent's datacenter | The datacenter to forward requests to.
 readConsistencyMode | `default` | Select between [Consul API consistency modes](https://www.consul.io/docs/agent/http.html) such as `default`, `stale` and `consistent` for reads.
 writeConsistencyMode | `default` | Select between [Consul API consistency modes](https://www.consul.io/docs/agent/http.html) such as `default`, `stale` and `consistent` for writes.
+failFast | `false` | If `false`, disable fail fast and failure accrual for Consul client. Should be changed to `true` if `host` is a load-balanced DNS name
```
Comment (adleong) [inline namerd/docs/storage.md]: The advice in the last sentence is a bit confusing.  Maybe say something about keeping it false if talking to consul via the local consul agent but setting to true if talking directly to a HA consul api server?

File: consul/src/main/scala/io/buoyant/consul/utils.scala (hunk 2 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: OK. I'd be very grateful if you could briefly explain what it gives us or point me to some doc about that.

File: consul/src/main/scala/io/buoyant/consul/utils.scala (hunk 1 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: Yeah, makes sense.

File: consul/src/main/scala/io/buoyant/consul/utils.scala (hunk 3 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: Oooops, sorry. Will fix.

File: CHANGES.md (hunk 4 above)
Comment (Ashald) [inline CHANGES.md]: Do we need an extra `*8 Breaking Chnage **` note? :)

File: namerd/docs/storage.md (hunk 5 above)
Comment (Ashald) [inline namerd/docs/storage.md]: Thanks!

File: CHANGES.md (hunk 4 above)
Comment (adleong) [inline CHANGES.md]: I'd say that's probably not necessary

File: consul/src/main/scala/io/buoyant/consul/utils.scala (hunk 2 above)
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/utils.scala]: http://docs.scala-lang.org/overviews/core/value-classes.html

**Comment by Ashald** [comment]
Pushed re-base to latest master + fixes according to review

**Comment by olix0r** [comment]
Can we find a more descriptive name than "utils"?

**Comment by Ashald** [comment]
@olix0r do you have something in mind? :)

File: consul/src/main/scala/io/buoyant/consul/SetHostFilter.scala [hunk 6]
```diff
@@ -1,8 +1,9 @@
 package io.buoyant.consul
 
+import java.net.InetSocketAddress
+
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/SetHostFilter.scala]: revert please.  linkerd style is all imports together alphasorted

Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/SetHostFilter.scala]: Fixed

**Comment by adleong** [comment]
:star: great! looks good to me!

**Pull request merged**